service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,372 @@
+
+        GiftCertificate foundCertificate = foundCertificateOpt.orElseThrow(() ->
+                new GiftCertificateNotFoundException(String.format("Can't find a certificate with id: %d", id)));
+
+        GiftCertificateDto giftCertificateDto = modelMapper.map(foundCertificate, GiftCertificateDto.class);
+        fillCertificateDtoWithTags(giftCertificateDto);
+        return giftCertificateDto;
+    }
+
+
+    /**
+     * This method set date fields for GiftCertificate entity and saves it in database. Also the method link passed tags with the GiftCertificate
+     * entity and, if the tag doesn't exist in db, creates it.
+     *
+     * @param certificate GiftCertificate entity without date fields and id for saving.
+     * @return GiftCertificateDto with id and tags.
+     * @since 1.0
+     */
+    @Override
+    @Transactional
+    public GiftCertificateDto save(GiftCertificateDto certificate) {
+
+        String certificateName = certificate.getName();
+        Optional<GiftCertificate> foundCertOpt = giftCertificateDao.getByName(certificateName);//check existence of GiftCertificate
+        // in db by name
+
+        if (foundCertOpt.isPresent()) {
+            throw new GiftCertificateWithSuchNameAlreadyExists(String.format("Gift certificate with name: %s already exits.",
+                    certificate.getName()));
+        }
+
+        GiftCertificate giftCertificate = modelMapper.map(certificate, GiftCertificate.class);
+        String currentTime = DateHelper.getNowAsString();
+        giftCertificate.setCreateDate(currentTime);
+        giftCertificate.setLastUpdateDate(currentTime); //setting creationDate and lastUpdateDate
+
+        GiftCertificate savedCertificate = giftCertificateDao.save(giftCertificate);
+
+        Long certificateId = savedCertificate.getId();
+
+        List<TagDto> tags = certificate.getTags(); //checking passed tags
+        if (tags != null) {
+            tags.forEach(tagDto -> {
+                addTagToCertificate(tagDto, certificateId);
+            });
+        }
+
+        GiftCertificateDto savedCertificateDto = modelMapper.map(savedCertificate, GiftCertificateDto.class);
+        fillCertificateDtoWithTags(savedCertificateDto);
+
+        return savedCertificateDto;
+    }
+
+
+    /**
+     * This method does full update for GiftCertificate entity with given id.
+     *
+     * @param certificateDto DTO contains field for updating GiftCertificate entity.
+     * @param certificateId  id of updatable GiftCertificate entity.
+     * @return GiftCertificateDto with id and tags.
+     * @throws GiftCertificateNotFoundException if GiftCertificate entity with given id doesn't exist in db.
+     * @since 1.0
+     */
+    @Override
+    @Transactional
+    public GiftCertificateDto update(GiftCertificateDto certificateDto, long certificateId) {
+
+        Optional<GiftCertificate> foundCertificateOpt = giftCertificateDao.getById(certificateId);
+
+        foundCertificateOpt.orElseThrow(() ->
```

Comment:
So, I was thinking about your approach, it looks good at first point, but there is one moment.
Am I right that your update method can return only one error message? you are throwing GiftCertificateNotFoundException and other logic won't be perfomed. So let's imagine the situation, when we have a bit entity 'Certificate' with 10-20 fields, and there can be a lot of validation/error messages. So your application will return only one error message each time, even if client had errors in other fields, which were missed after throwing GiftCertificateNotFoundException. Instead of this, you can fully validate your 'Certificate' DTO and return List<ErrorMessages> - your application will return list of error messages and client will fix all errors at once, instead of  fixing one error and getting one error message at a time.

Reply:
Validation exception messages now forms in dynamically depends on incorrect fields in the passed DTO.

commit 3c82372